src/ints/mouse.cpp

```diff
@@ -18,1549 +18,932 @@
-		DrawCursorText();
-		return;
-	}
-
-	// Check video page. Seems to be ignored for text mode.
-	// hence the text mode handled above this
-	// >>> removed because BIOS page is not actual page in some cases, e.g.
-	// QQP games
-	//    if (real_readb(BIOSMEM_SEG,BIOSMEM_CURRENT_PAGE)!=mouse.page) return;
-
-	// Check if cursor in update region
-	/*    if ((POS_X >= mouse.updateRegion_x[0]) && (POS_X <=
-	   mouse.updateRegion_x[1]) && (POS_Y >= mouse.updateRegion_y[0]) &&
-	   (POS_Y <= mouse.updateRegion_y[1])) { if (CurMode->type==M_TEXT16)
-	            RestoreCursorBackgroundText();
-	        else
-	            RestoreCursorBackground();
-	        mouse.shown--;
-	        return;
-	    }
-	   */ /*Not sure yet what to do update region should be set to ??? */
-
-	// Get Clipping ranges
-
-	mouse.clipx = (int16_t)((Bits)CurMode->swidth - 1); /* Get from bios ? */
-	mouse.clipy = (int16_t)((Bits)CurMode->sheight - 1);
-
-	/* might be vidmode == 0x13?2:1 */
-	int16_t xratio = 640;
-	if (CurMode->swidth > 0)
-		xratio /= CurMode->swidth;
-	if (xratio == 0)
-		xratio = 1;
-
-	RestoreCursorBackground();
-
-	SaveVgaRegisters();
-
-	// Save Background
-	int16_t x, y;
-	uint16_t addx1, addx2, addy;
-	uint16_t dataPos = 0;
-	int16_t x1       = POS_X / xratio - mouse.hotx;
-	int16_t y1       = POS_Y - mouse.hoty;
-	int16_t x2       = x1 + CURSORX - 1;
-	int16_t y2       = y1 + CURSORY - 1;
-
-	ClipCursorArea(x1, x2, y1, y2, addx1, addx2, addy);
-
-	dataPos = addy * CURSORX;
-	for (y = y1; y <= y2; y++) {
-		dataPos += addx1;
-		for (x = x1; x <= x2; x++) {
-			INT10_GetPixel(x, y, mouse.page, &mouse.backData[dataPos++]);
-		};
-		dataPos += addx2;
-	};
-	mouse.background = true;
-	mouse.backposx   = POS_X / xratio - mouse.hotx;
-	mouse.backposy   = POS_Y - mouse.hoty;
-
-	// Draw Mousecursor
-	dataPos = addy * CURSORX;
-	for (y = y1; y <= y2; y++) {
-		uint16_t scMask = mouse.screenMask[addy + y - y1];
-		uint16_t cuMask = mouse.cursorMask[addy + y - y1];
-		if (addx1 > 0) {
-			scMask <<= addx1;
-			cuMask <<= addx1;
-			dataPos += addx1;
-		};
-		for (x = x1; x <= x2; x++) {
-			uint8_t pixel = 0;
-			// ScreenMask
-			if (scMask & HIGHESTBIT)
-				pixel = mouse.backData[dataPos];
-			scMask <<= 1;
-			// CursorMask
-			if (cuMask & HIGHESTBIT)
-				pixel = pixel ^ 0x0F;
-			cuMask <<= 1;
-			// Set Pixel
-			INT10_PutPixel(x, y, mouse.page, pixel);
-			dataPos++;
-		};
-		dataPos += addx2;
-	};
-	RestoreVgaRegisters();
-}
+} MouseEvent;
 
-static void CursorMoved(float xrel, float yrel, float x, float y, bool emulate)
+static class MouseQueue final {
+public:
+    MouseQueue();
+
+    void AddEvent(MouseEvent &event);
+    void FetchEvent(MouseEvent &event);
+    void ClearEventsDOS();
+    void StartTimerIfNeeded();
+    void Tick();
+
+    uint8_t start_delay_ps2     = 5; // for PS/2 events, in milliseconds
+    uint8_t start_delay_dos_btn = 1; // for DOS button events
+    uint8_t start_delay_dos_mov = 5; // for DOS move/wheel events
```

Comment:
I think I'd like `button` and `move` better; `btn` and especially `mov` remind me of assembler mnemonics 😄

Reply:
I'd suggest `_ms` suffixes for all these.

Reply:
Done.